common/src/main/java/org/conscrypt/HpkeImpl.java

```diff
@@ -0,0 +1,141 @@
+package org.conscrypt;
+
+import static org.conscrypt.HpkeSuite.AEAD_AES_128_GCM;
+import static org.conscrypt.HpkeSuite.AEAD_AES_256_GCM;
+import static org.conscrypt.HpkeSuite.AEAD_CHACHA20POLY1305;
+import static org.conscrypt.HpkeSuite.KDF_HKDF_SHA256;
+import static org.conscrypt.HpkeSuite.KEM_DHKEM_X25519_HKDF_SHA256;
+
+import java.security.InvalidKeyException;
+import java.security.PrivateKey;
+import java.security.Provider;
+import java.security.PublicKey;
+
+public class HpkeImpl implements HpkeSpi {
+  private final Provider provider;
+  private final HpkeSuite hpkeSuite;
+
+  private NativeRef.EVP_HPKE_CTX ctx;
+  private byte[] enc = null;
+
+
+  public HpkeImpl(Object arg, HpkeSuite hpkeSuite) {
+    if (arg instanceof Provider) {
+      provider = (Provider) arg;
+    } else {
+      provider = null;
+    }
+    this.hpkeSuite = hpkeSuite;
+  }
+
+  @Override
+  public void engineInitSender(int mode, PublicKey publicKey, byte[] info, byte[] sKe) throws InvalidKeyException {
+    if (publicKey == null) {
+      throw new InvalidKeyException("null key");
+    }
+    if (mode != HpkeContextSender.MODE_BASE) {
+      throw new UnsupportedOperationException("Unsupported mode " + mode);
+    }
+    final byte[] pk = hpkeSuite.getKem().validatePublicKeyTypeAndGetRawKey(publicKey);
+    checkNotInitialised();
+
+    final Object[] result = (sKe == null)
+        ? NativeCrypto.EVP_HPKE_CTX_setup_sender(hpkeSuite.getKem().getId(),
+        hpkeSuite.getKdf().getId(), hpkeSuite.getAead().getId(), pk, info)
+        : NativeCrypto.EVP_HPKE_CTX_setup_sender_with_seed_for_testing(hpkeSuite.getKem().getId(),
+        hpkeSuite.getKdf().getId(), hpkeSuite.getAead().getId(), pk, info, sKe);
+    ctx = (NativeRef.EVP_HPKE_CTX) result[0];
+    enc = (byte[]) result[1];
+  }
+
+  @Override
+  public void engineInitRecipient(int mode, byte[] enc, PrivateKey key, byte[] info) throws InvalidKeyException {
```

Comment:
Could we add the validation back on the `enc` length? Or is the intention that `EVP_HPKE_CTX_setup_recipient ` takes care of it?

Reply:
Oops, looks like a cut and paste error, will go check.